src/IO/RemoteReadBufferCache.cpp

```diff
@@ -0,0 +1,585 @@
+        if (cache_iter->second.cache_controller->getLastModificationTimestamp() != last_modification_timestamp)
+        {
+            LOG_TRACE(log,
+                "remote file has been updated. " + remote_path + ":" + std::to_string(cache_iter->second.cache_controller->getLastModificationTimestamp()) + "->"
+                    + std::to_string(last_modification_timestamp));
+            cache_iter->second.cache_controller->markInvalid();
+        }
+        else
+        {
+            // move the key to the list end
+            keys.splice(keys.end(), keys, cache_iter->second.key_iterator);
+            return {std::make_shared<LocalCachedFileReader>(cache_iter->second.cache_controller.get(), file_size), RemoteReadBufferCacheError::OK};
+        }
+    }
+
+    auto clear_ret = clearLocalCache();
+    cache_iter = caches.find(local_path);
+    if (cache_iter != caches.end())
+    {
+        if (cache_iter->second.cache_controller->isValid())
+        {
+            // move the key to the list end, this case should not happen?
+            keys.splice(keys.end(), keys, cache_iter->second.key_iterator);
+            return {std::make_shared<LocalCachedFileReader>(cache_iter->second.cache_controller.get(), file_size), RemoteReadBufferCacheError::OK};
+        }
+        else
+        {
+            // maybe someone is holding this file
+            return {nullptr, RemoteReadBufferCacheError::FILE_INVALID};
+        }
+    }
+
+    // reach the disk capacity limit
+    if (!clear_ret)
+    {
+        LOG_ERROR(log, "local cache is full, return nullptr");
+        return {nullptr, RemoteReadBufferCacheError::DISK_FULL};
+    }
+
+    std::filesystem::create_directories(local_path);
+
+    auto callback = [this](RemoteCacheController * cntrl) { this->total_size += cntrl->size(); };
+    auto cache_cntrl = std::make_shared<RemoteCacheController>(remote_file_meta, local_path, local_cache_bytes_read_before_flush, readbuffer, callback);
+    CacheCell cc;
+    cc.cache_controller = cache_cntrl;
+    cc.key_iterator = keys.insert(keys.end(), local_path);
+    caches[local_path] = cc;
+
+    return {std::make_shared<LocalCachedFileReader>(cache_cntrl.get(), file_size), RemoteReadBufferCacheError::OK};
+}
+
+bool RemoteReadBufferCache::clearLocalCache()
+{
+    for (auto it = keys.begin(); it != keys.end();)
+    {
+        auto cache_it = caches.find(*it);
+        auto cntrl = cache_it->second.cache_controller;
+        if (!cntrl->isValid() && cntrl->closable())
+        {
+            LOG_TRACE(log, "clear invalid cache: " + *it);
+            total_size = total_size > cache_it->second.cache_controller->size() ? total_size - cache_it->second.cache_controller->size() : 0;
+            cntrl->close();
+            it = keys.erase(it);
+            caches.erase(cache_it);
+        }
+        else
+            it++;
+    }
+    // clear closable cache from the list head
+    for (auto it = keys.begin(); it != keys.end();)
```

Comment:
Looks like this two loops can be easily put into one.

Reply:
done